proposals/compound-extensions.md

```diff
@@ -0,0 +1,380 @@
+       }
+       ```
+
+       ```kotlin
+       fun main() = runBlocking {
+         numbersFrom(2).filter(this) { it % 2 }.forEach { println(it) }
+       }
+       ```
+
+    With this proposal the developer could introduce use a more fluent `filter`,
+
+    ```kotlin
+    fun <T> CoroutineScope.ReceiverChannel<T>.filter(
+        predicate: (value: T) -> Boolean) = produce<T> {
+      for (value in this) if (predicate) send(value)
+    }
+    ```
+
+    ```kotlin
+    fun main() = runBlocking {
+        numbersFrom(2).filter { it % 2 }.forEach { println(it) }
+    }
+    ```
+
+* Generalizing local extensions
+
+  Extensions in local scope are difficult to extract. Consider a simplified HTML DSL
+  such as:
+
+    ```kotlin
+    open class HtmlDsl() { fun write(s: String) {  } fun escape(s: String) { } }
+    class Html(): HtmlDsl() {}
+    class Body(): HtmlDsl() { }
+
+    fun html(block: Html.() -> Unit) {
+        with (Html()) {
+            write("<html>")
+            block()
+            write("</html>")
+        }
+    }
+
+    fun Html.body(block: Body.() -> Unit) {
+        write("<body>")
+        Body().block()
+        write("<body>")
+    }
+
+    fun Body.div(block: Body.() -> Unit) {
+        write("<div>")
+        block()
+        write("</div>")
+    }
+
+    fun Body.span(block: Body.() -> Unit) {
+        write("<span>")
+        block()
+        write("</span>")
+    }
+    ```
+
+   To emit string content it would be convenient to overload the unary `+` operator to
+   mean emit text with html escapes. The developer cannot use extension methods for this
+   to be generally available and must add the operator to the correct classes directly
+   such as,
+
+   ```kotlin
+   class Body(): HtmlDsl() {
+       operator fun String.unaryPlus(s: String) = escape(s)
+   }
+   ```
+
+   In a local scope the developer can declare extension operators for special purposes
+   such as introducing a unary `-` operator to underline text,
+
+   ```kotlin
+   html {
+        body {
+            operator fun String.unaryMinus() {
+                underline {
+                    +this@unaryMinus
+                }
+            }
+            div {
+                -"Customer"
+                customer(customer)
+
+                -"Orders"
+                orders(orders)
+            }
+        }
+    }
+    ```
+
+    However, the developer is required to modify `Body` class to allow the unary `-` to be
+    used outside its current local scope. This is problematic if the HTML DSL is a third-party
+    library.
+
+    Compound extensions allows the original unary `+` operator to be introduced as an extension
+    function such as,
+
+    ```kotlin
+    operator fun Body.String.unaryPlus() = escape(this)
+    ```
+
+    therefore allowing unary `-` to be provided generally by extension of `Body`,
+
+    ```kotlin
+    operator fun Body.String.unaryMinus() = underline { +this@unaryMinus }
+    ```
+
+## Proposal
```

Comment:
This proposal does not address limitations regarding two receiver types being related.  Would it be possible to define an extension with two receivers of the same type, and if so, how would they be disambiguated?

Reply:
This question is answered in the  discussion starting at https://github.com/Kotlin/KEEP/pull/176#issuecomment-460147227